### PR TITLE
[batch] cookies, headers, authentication

### PR DESCRIPTION
**Pull request by danking**
cc: @akotlar @cseed 

I whipped this up today. It modifies the batch api to accept user-configurable cookies and headers. The `BatchClient` looks for a token in `~/.hail/batch/token` or at a user-provided path. There's a whitelist of authorized users. If you can talk directly to batch, you can spoof the `User` header. This makes testing easy. For this approach to be secure, we need to ensure:
- only `gateway` may send HTTP requests to `batch`
- `gateway` does not forward any "secure" headers (currently, just `User`)

### CODE REVIEW

**Comment by danking** [comment]
Lemme know what you think of this approach. @jigold let's not approve until there's some consensus.

File: batch/batch/client.py [hunk 1]
```diff
@@ -91,19 +92,27 @@ def wait(self):
...
 
 class BatchClient:
-    def __init__(self, url=None, api=api.DEFAULT_API):
+    def __init__(self, url=None, timeout=None, token_file=None, headers=None):
         if not url:
             url = 'http://batch.default'
         self.url = url
-        self.api = api
+        if token_file is None:
+            token_file = os.path.expanduser('~/.hail/batch/token')
```
Comment (cseed) [inline batch/batch/client.py]: The authentication is Hail services wide, so I think just `~/.hail/token`.

File: batch/batch/server/server.py [hunk 2]
```diff
@@ -417,7 +417,39 @@ def to_json(self):
...
+    'dking@broadinstitute.org',
+    'akotlar@broadinstitute.org',
+    'jigold@broadinstitute.org',
+    'ci'
+]
+
+
+def authorized_users_only(fun):
+    def wrapped(*args, **kwargs):
+        if not request.headers.get('User', '') in authorized_users:
```
Comment (cseed) [inline batch/batch/server/server.py]: I think our custom headers should be prefixed by `Hail-` and we should use Capitals-With-Dashes which seems standard.  So, `Hail-Internal-Request-None`?

File: batch/test/test_dag.py [hunk 3]
```diff
@@ -370,3 +371,24 @@ def test_input_dependency(client):
...
+        (requests.delete, '/jobs/0/delete'),
+        (requests.post, '/jobs/0/cancel'),
+        (requests.post, '/batches/create'),
+        (requests.get, '/batches/0'),
+        (requests.delete, '/batches/0/delete'),
+        (requests.post, '/batches/0/close'),
+        (requests.post, '/pod_changed'),
+        (requests.post, '/refresh_k8s_state'),
+        (requests.get, '/recent')]
+    for endpoint in endpoints:
```
Comment (cseed) [inline batch/test/test_dag.py]: I prefer `for f, path in endpoints: ...`.

File: batch/test/test_batch.py [hunk 4]
```diff
@@ -11,7 +11,9 @@
 
 class Test(unittest.TestCase):
     def setUp(self):
-        self.batch = batch.client.BatchClient(url=os.environ.get('BATCH_URL'))
+        self.batch = batch.client.BatchClient(
+            url=os.environ.get('BATCH_URL'),
+            headers={'User': 'dking@broadinstitute.org'})
```
Comment (cseed) [inline batch/test/test_batch.py]: I don't think we should have your email in there.

File: batch/batch/api.py [hunk 5]
```diff
@@ -35,32 +43,51 @@ def create_job(self, url, spec, attributes, batch_id, callback, parent_ids,
         if copy_service_account_name:
             doc['copy_service_account_name'] = copy_service_account_name
 
-        response = requests.post(url + '/jobs/create', json=doc, timeout=self.timeout)
+        response = requests.post(url + '/jobs/create',
```
Comment (cseed) [inline batch/batch/api.py]: might be easier to have `self.post(...)` with defaults.

File: batch/batch/server/server.py [hunk 6]
```diff
@@ -417,7 +417,39 @@ def to_json(self):
 log.info(f'app.root_path = {app.root_path}')
 
 
+authorized_users = [
+    'konradk@broadinstitute.org',
+    'labbott@broadinstitute.org',
+    'dking@broadinstitute.org',
+    'akotlar@broadinstitute.org',
+    'jigold@broadinstitute.org',
+    'ci'
```
Comment (cseed) [inline batch/batch/server/server.py]: @akotlar How will our service accounts look?

File: batch/batch/server/server.py [hunk 7]
```diff
@@ -417,7 +417,39 @@ def to_json(self):
 log.info(f'app.root_path = {app.root_path}')
 
 
+authorized_users = [
+    'konradk@broadinstitute.org',
```
Comment (cseed) [inline batch/batch/server/server.py]: I think auth0 user ids are prefixed with the provider but I forget the details.  @akotlar?

Comment (akotlar) [reply batch/batch/server/server.py]: auth0 ids are prefixed as you say, except `google-oauth2|id` instead of `google|id`: https://auth0.com/docs/api-auth/tutorials/adoption/scope-custom-claims
- some more detail: https://community.auth0.com/t/sub-claim-format-kubernetes-restricted-set/21996

File: batch/batch/server/server.py (hunk 6 above)
Comment (akotlar) [reply batch/batch/server/server.py]: Plan was to name all service accounts by the user id (`sub` claim). This needs to be of course trasnfrormed because of the restricted dchartersets both google cloud and Kubernetes employ. Kube names (253 chars, `[a-zA-Z0-9-]*`: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/)

Google service accounts: Name is auto-generated, using a prefix that you supply, which I intended to be the same `sub` claim (user id). For a name `google-oauth2-kotlar-test`

```
{'name': 'projects/hail-vdc/serviceAccounts/google-oauth2-kotlar-test@hail-vdc.iam.gserviceaccount.com', 'projectId': 'hail-vdc', 'uniqueId': '101901669302330053293', 'email': 'google-oauth2-kotlar-test@hail-vdc.iam.gserviceaccount.com', 'etag': 'MDEwMjE5MjA=', 'oauth2ClientId': '101901669302330053293'}
```

Google Cloud service account characterset is slightly more restrictive than Kubernetes: (` [a-z][a-z\d\-]*[a-z\d]`) (no '.' allowed):

```
googleapiclient.errors.HttpError: <HttpError 400 when requesting https://iam.googleapis.com/v1/projects/hail-vdc/serviceAccounts?alt=json returned "google-oauth2|kotlar-test does not match [a-z][a-z\d\-]*[a-z\d].">
```

Comment (akotlar) [reply batch/batch/server/server.py]: Transformation can continue to be a hash of the user id, as done by notebook2 now (sha224, whose hex digest < 63 chars, which is the largest a label allows), or, if we trust that the user id is truly <provider>|<[a-Z0-9-]+, just replace the pipe with a dash. However, when I inquired about the guarantees, the answer I got suggested there is no guarantee: https://community.auth0.com/t/sub-claim-format-kubernetes-restricted-set/21996

...and the OpenID Connect spec that Auth0 follows specifies ASCII, which obviously means relying on Auth0 to use alphanumerics is fragile.

Comment (akotlar) [reply batch/batch/server/server.py]: A note on id uniqueness, stability: https://openid.net/specs/openid-connect-core-1_0.html#ClaimStability

File: batch/batch/api.py (hunk 5 above)
Comment (danking) [reply batch/batch/api.py]: done

File: batch/batch/client.py (hunk 1 above)
Comment (danking) [reply batch/batch/client.py]: done

File: batch/batch/server/server.py (hunk 2 above)
Comment (danking) [reply batch/batch/server/server.py]: done

File: batch/batch/server/server.py (hunk 6 above)
Comment (danking) [reply batch/batch/server/server.py]: can we keep a mapping in a db somewhere from auth0 name to an int64 and use that everywhere as the user id? makes debugging annoying, but there seems to be a lot of issues that disappear if we use an id.

File: batch/test/test_batch.py (hunk 4 above)
Comment (danking) [reply batch/test/test_batch.py]: switched to `ci`

File: batch/test/test_dag.py (hunk 3 above)
Comment (danking) [reply batch/test/test_dag.py]: done, also moved to test_batch since it's not a dag test.

File: batch/batch/server/server.py (hunk 6 above)
Comment (cseed) [reply batch/batch/server/server.py]: @danking agreed.

Comment (akotlar) [reply batch/batch/server/server.py]: The solution I'm using at the moment is for service names.

user-<shortuuid> (22 character/byte uuid)

This id will get stored in the user's id token, which provides the mapping. We can of course maintain our own user db, but if we have the ability to store custom claims, is that not sufficient?

Comment (cseed) [reply batch/batch/server/server.py]: I think he meant the same db that stores the other claims like bucket and service accounts.

 > The solution I'm using at the moment for service names

You mean k8s service names?  For maximum confusion, there are three kinds of service accounts: k8s service accounts, GCP service accounts, and Hail service accounts like ci.  Hail service accounts use Hail services like Hail users do, but are robots instead of users.  Hail service accounts need all the usual Hail users claims, like buckets and k8s and GCP service accounts, etc.

Comment (akotlar) [reply batch/batch/server/server.py]: For the initial approach, I am doing effectively:

```
create(username):
    sa_name = f'user-{shortuuid.uuid()}'

    out['sa_name'] = sa_name
    
    # Store username in 'displayName' property
    gsa_response = make_google_service_account(sa_name, username)
    out['gsa_unique_id'] = gsa_response['uniqueId']
    
    # Stores username in 'username' metadata
    ksa_response = make_kube_service_account(sa_name, username)
    
    out['ksa_name'] = sa_name

    out['bucket'] = create_bucket(sa_name)
```

The Google IAM response: 
```
{'name': 'projects/hail-vdc/serviceAccounts/user-14gwognesw8rw56xb3je73xgh@hail-vdc.iam.gserviceaccount.com', 'projectId': 'hail-vdc', 'uniqueId': '117029443340792843240', 'email': 'user-14gwognesw8rw56xb3je73xgh@hail-vdc.iam.gserviceaccount.com', 'displayName': 'akotlar', 'etag': 'MDEwMjE5MjA=', 'oauth2ClientId': '117029443340792843240'}
```

The gcloud output
```
$ gcloud iam service-accounts list
NAME                                    EMAIL
akotlar                                 user-02p21ibcb1axsokf1wp3vtcpd@hail-vdc.iam.gserviceaccount.com
```

The kube response
```
{'api_version': 'v1',
 'automount_service_account_token': None,
 'image_pull_secrets': None,
 'kind': 'ServiceAccount',
 'metadata': {'annotations': {'kubernetes.io/service-account.name': 'user-1a4zr260kqx8256mcg6mn3ue6',
                              'username': 'akotlar'},
              'cluster_name': None,
              'creation_timestamp': datetime.datetime(2019, 3, 15, 16, 56, 54, tzinfo=tzutc()),
              'deletion_grace_period_seconds': None,
              'deletion_timestamp': None,
              'finalizers': None,
              'generate_name': None,
              'generation': None,
              'initializers': None,
              'labels': None,
              'name': 'user-1a4zr260kqx8256mcg6mn3ue6',
              'namespace': 'default',
              'owner_references': None,
              'resource_version': '35974418',
              'self_link': '/api/v1/namespaces/default/serviceaccounts/user-1a4zr260kqx8256mcg6mn3ue6',
              'uid': '563c1dbe-4743-11e9-bcd4-42010a8000c9'},
 'secrets': None}
```

Comment (akotlar) [reply batch/batch/server/server.py]: Happy to go a different direction :)

Comment (danking) [reply batch/batch/server/server.py]: user-uuid sounds great.

I feel a bit weird about using a SHA (which I think is where the uuid comes from?) because it seems like we should be able to deterministically hand out int64's to users. I'm mostly ok with the uuid approach if it's substantially simpler.

**Comment by danking** [comment]
I ported pipeline to new batch, and in doing so I ripped out `copy_service_account_name` and just replaced its uses with `user`.

File: batch/batch/server/server.py [hunk 8]
```diff
@@ -426,14 +432,49 @@ def to_json(self):
...
+]
+
+
+def authorized_users_only(fun):
+    def wrapped(*args, **kwargs):
+        user = request.headers.get('Hail-User', '')
+        if user not in authorized_users:
+            return '403 UNAUTHORIZED', 403
+        if 'user' in fun.__code__.co_varnames:
+            return fun(*args, user=user, **kwargs)
```
Comment (jigold) [inline batch/batch/server/server.py]: That's clever!

File: pipeline/pipeline/backend.py [hunk 9]
```diff
@@ -143,48 +143,35 @@ def run(self, pipeline, dry_run, verbose, delete_scratch_on_exit):  # pylint: di
...
                     assert isinstance(r, ResourceGroup)
                     return [x for _, rf in r._resources.items() for x in copy_output(rf)]
 
             copy_task_inputs = [x for r in task._inputs for x in copy_input(r)]
             copy_task_outputs = [x for r in task._outputs for x in copy_output(r)]
 
-            local_dirs_needed = [local_tmpdir,
-                                 local_tmpdir + '/inputs/',
-                                 local_tmpdir + f'/{task._uid}/']
+            local_dirs_needed = [local_tmpdir + f'/{task._uid}/']
```
Comment (jigold) [inline pipeline/pipeline/backend.py]: I don't think we need `local_dirs_needed` if gustil does copying now.

File: pipeline/pipeline/backend.py [hunk 10]
```diff
@@ -143,48 +143,35 @@ def run(self, pipeline, dry_run, verbose, delete_scratch_on_exit):  # pylint: di
...
             local_dirs_needed += [local_tmpdir + f'/{dep._uid}/' for dep in task._dependencies]
             make_local_tmpdir = [f'mkdir -p {dir}' for dir in local_dirs_needed]
```
Comment (jigold) [inline pipeline/pipeline/backend.py]: Can probably delete this too

File: pipeline/pipeline/backend.py [hunk 11]
```diff
@@ -208,54 +195,28 @@ def copy_output(r):
...
+                                 output_files=copy_task_outputs)
             n_jobs_submitted += 1
 
             task_to_job_mapping[task] = j
             job_id_to_command[j.id] = defs + cmd
             if verbose:
                 print(f"Submitted Job {j.id} with command: {defs + cmd}")
 
-        def write_pipeline_outputs(r, dest):
-            if isinstance(r, InputResourceFile):
```
Comment (jigold) [inline pipeline/pipeline/backend.py]: I thought I had this separate for a reason from the copying above. I need to think further if this is kosher.

File: pipeline/test/test_pipeline.py [hunk 12]
```diff
@@ -245,7 +245,8 @@ def test_add_extension_input_resource_file(self):
 
 class BatchTests(unittest.TestCase):
     def pipeline(self):
-        return Pipeline(backend=BatchBackend(os.environ.get('BATCH_URL')),
+        return Pipeline(backend=BatchBackend(os.environ.get('BATCH_URL'),
+                                             user='pipeline-test-0-1--hail-is'),
```
Comment (jigold) [inline pipeline/test/test_pipeline.py]: Formatting!

File: batch/batch/server/server.py [hunk 13]
```diff
@@ -426,14 +432,49 @@ def to_json(self):
 log.info(f'app.root_path = {app.root_path}')
 
 
+authorized_users = [
+    'ci',
+    'pipeline-test-0-1--hail-is',
+    'google-oauth2|konradk@broadinstitute.org',
+    'google-oauth2|labbott@broadinstitute.org',
+    'google-oauth2|dking@broadinstitute.org',
+    'google-oauth2|akotlar@broadinstitute.org',
```
[the remark this answers is not in the document]
Comment (cseed) [reply batch/batch/server/server.py]: Me, too!  Me, too!

File: pipeline/test/test_pipeline.py (hunk 12 above)
Comment (danking) [reply pipeline/test/test_pipeline.py]: I think this is actually right, `user` is a keyword argument to `BatchBackend` and the close-paren ends the `BatchBackend` constructor.

File: pipeline/pipeline/backend.py (hunk 11 above)
Comment (danking) [reply pipeline/pipeline/backend.py]: I changed the semantics slightly. Now, [I eagerly copy the results out](https://github.com/hail-is/hail/pull/5605/files/35cf6f149ade56d790deb0ebf65d516110b42747#diff-2e8d9bac670a9022b6a3bedd33315b56R163) when a job is finished. You waited for the entire pipeline to finish before copying all pipeline outputs out.

File: pipeline/test/test_pipeline.py (hunk 12 above)
Comment (jigold) [reply pipeline/test/test_pipeline.py]: ok.

File: batch/batch/server/server.py [hunk 14]
```diff
@@ -442,14 +448,50 @@ def to_dict(self):
...
+    'google-oauth2|jigold@broadinstitute.org',
+    'google-oauth2|cseed@broadinstitute.org'
+]
+
+
+def authorized_users_only(fun):
+    def wrapped(request, *args, **kwargs):
+        user = request.headers.get('Hail-User', '')
```
Comment (jigold) [inline batch/batch/server/server.py]: Where does the cookie / authorization verification come into this?

File: batch/batch/client.py [hunk 15]
```diff
@@ -200,9 +207,6 @@ def _get_batch(self, batch_id):
     def _close_batch(self, batch_id):
         return self.api.close_batch(self.url, batch_id)
 
-    def _refresh_k8s_state(self):
```
Comment (jigold) [inline batch/batch/client.py]: Why did you delete this?

File: pipeline/pipeline/backend.py [hunk 16]
```diff
@@ -133,7 +133,7 @@ def run(self, pipeline, dry_run, verbose, delete_scratch_on_exit):  # pylint: di
         subdir_name = 'pipeline-{}'.format(uuid.uuid4().hex[:12])
```
Comment (jigold) [inline pipeline/pipeline/backend.py]: Ultimately, we want to use the scratch bucket specified by the user -- correct?

File: pipeline/pipeline/backend.py [hunk 17]
```diff
@@ -143,48 +143,31 @@ def run(self, pipeline, dry_run, verbose, delete_scratch_on_exit):  # pylint: di
...
                     assert isinstance(r, ResourceGroup)
                     return [x for _, rf in r._resources.items() for x in copy_input(rf)]
 
             def copy_output(r):
                 assert r._source == task  # pylint: disable-msg=W0640
                 if isinstance(r, TaskResourceFile):
-                    return [f'gsutil cp {r._get_path(local_tmpdir)} {r._get_path(remote_tmpdir)}']
+                    return [
+                        (r._get_path(local_tmpdir), r._get_path(remote_tmpdir)),
+                        *[(r._get_path(local_tmpdir), output_path) for output_path in r._output_paths]]
```
Comment (jigold) [inline pipeline/pipeline/backend.py]: This changes the behavior of my original code. You can write out an input resource file, but your code won't write them out because it's only looking at the task outputs. That's why I had the code at the end that went through the entire map of a pipeline's resources and wrote everything out at once. Perhaps adding an additional data structure that looks for input files will solve this. I also had code to handle a resource group and writing the file with the extension being the identifier of the resource group's file. Example `p.write_output(t.bfile, 'gs://example') should write out gs://example.bed, gs://example.bim, and gs://example.fam`

Also, should probably add asserts that source != destination when copying in Batch.

File: batch/batch/client.py (hunk 15 above)
Comment (danking) [reply batch/batch/client.py]: unused

File: batch/batch/server/server.py (hunk 14 above)
Comment (danking) [reply batch/batch/server/server.py]: I added JWT decoding, that ensures its a valid cookie issued by us

File: pipeline/pipeline/backend.py (hunk 16 above)
Comment (danking) [reply pipeline/pipeline/backend.py]: yes

**Comment by danking** [comment]
closing while I iterate on this

**Comment by jigold** [comment]
@danking What is your plan for this PR? Can I do anything to help?
